### PR TITLE
DOC: correction to numpy.pad docstring

### PR DESCRIPTION
**Pull request by grlee77**
When a custom function is provided to `numpy.pad`, the docstring states that this function should return a rank 1 vector.  However, it actually seems that it is required that the rank 1 vector input be modified in-place and it is not necessary to return an array.

I also updated a test case to make sure it is consistent with the modified description proposed here.

Two slight modifications to the example from the docstring illustrate this behavior.

1.) Padding fails if a copy of `vector` is made internally:

```Python
def pad_with(vector, pad_width, iaxis, kwargs):
    pad_value = kwargs.get('padder', 10)
    vector = vector.copy()
    vector[:pad_width[0]] = pad_value
    vector[-pad_width[1]:] = pad_value
    return vector
a = np.arange(6)
a = a.reshape((2, 3))
np.pad(a, 2, pad_with)
```
> Out:
   array([[0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 1, 2, 0, 0],
       [0, 0, 3, 4, 5, 0, 0],
       [0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0]])

2.) Padding works if modification is done in-place, even without any return value:
```Python
def pad_with(vector, pad_width, iaxis, kwargs):
    pad_value = kwargs.get('padder', 10)
    vector[:pad_width[0]] = pad_value
    vector[-pad_width[1]:] = pad_value
    return
a = np.arange(6)
a = a.reshape((2, 3))
np.pad(a, 2, pad_with)
```
> Out:
array([[10, 10, 10, 10, 10, 10, 10],
       [10, 10, 10, 10, 10, 10, 10],
       [10, 10,  0,  1,  2, 10, 10],
       [10, 10,  3,  4,  5, 10, 10],
       [10, 10, 10, 10, 10, 10, 10],
       [10, 10, 10, 10, 10, 10, 10]])

### CODE REVIEW

**Comment by mattip** [comment]
LGTM, the docstring check passes so it seems the documentation fix is correct. Is there a reason the design is to modify the vector in-place? Is it more efficient that way? Perhaps the docstring could hint at the motivation for the (in my opinion) non-standard design.

**Comment by grlee77** [comment]
> Is there a reason the design is to modify the vector in-place?

I am not entirely sure if that is intentional, but the relevant usage of the provided function occurs with `apply_along_axis`:

https://github.com/numpy/numpy/blob/17b31fdfaf7cca752de9d3590f5caeb08b3f5c18/numpy/lib/arraypad.py#L1223-L1236

In that section of code, the output of `apply_along_axis` is never assigned, so it must be relying on in-place modification of `newmat`. I assume this is for reason of efficiency, but am not sure. That section of code does not seem to have been modified in years.

I guess now the question is whether to regard this as an error in documentation or as a bug in `np.pad`. I think if the output of `np.apply_along_axis` were assigned back to `newmat` within the loop above then perhaps the existing docstring could be left as-is.

I came across the issue when working on PyWavelets/pywt#478. I was surprised to find that it was necessary to avoid an internal copy, which made me suspect that the return value was not being used.

**Comment by grlee77** [comment]
I started looking at `np.apply_along_axis` and it seems that it is still going to be allocating a new output array, even if it is never being used. In that case, I am not sure there is any benefit to relying on it doing the modification in-place.

**Comment by grlee77** [comment]
I did some preliminary testing and did not see any performance difference for padding a relatively large array of size (256, 256, 256) by 2 with or without assignment to `newmat` in the loop, so I don't think it would be a problem to fix the behaviour to match the current docstring.

However, if we do keep the requirement for in-place modification, it is possible to get a substantial performance improvement (I saw around 30-40% for my test case) if we continue to rely on in-place modification, but replace calls to `np.apply_along_axis` with calls to a locally defined version of that function that assumes in-place operation.

i.e. using `_apply_along_axis_inplace` as defined below (basically just an excerpt from the existing apply_along_axis, omitting the buffer allocation and assignment code):

```Python
def _apply_along_axis_inplace(func1d, axis, arr, *args, **kwargs):
    """
    Apply a function to 1-D slices along the given axis.

    This is like apply_along_axis, but relies on the modification to arr being
    done in-place.
    """
    # handle negative axes
    arr = asanyarray(arr)
    nd = arr.ndim
    axis = normalize_axis_index(axis, nd)

    # arr, with the iteration axis at the end
    in_dims = list(range(nd))
    inarr_view = transpose(arr, in_dims[:axis] + in_dims[axis+1:] + [axis])

    # compute indices for the iteration axes, and append a trailing ellipsis to
    # prevent 0d arrays decaying to scalars, which fixes gh-8642
    inds = ndindex(inarr_view.shape[:-1])
    inds = (ind + (Ellipsis,) for ind in inds)

    # invoke the function on the first item
    try:
        ind0 = next(inds)
    except StopIteration:
        raise ValueError('Cannot apply_along_axis when any iteration dimensions are 0')
    func1d(inarr_view[ind0], *args, **kwargs)
    for ind in inds:
        func1d(inarr_view[ind], *args, **kwargs)
    return
```
I suspect use of such user-defined functions is relatively rare, though, so I'm not sure it is worth optimizing this case?

**Comment by grlee77** [comment]
@lagru: I know you have been working on reducing memory allocations in `numpy.pad`. Do you have any opinion on whether we should update the docstring as proposed here (leaving open the possibility to optimize as suggested in the prior comment) or if we should just fix the behaviour so that it matches the current docstring?

**Comment by lagru** [comment]
First of, I have largely ignored that part of the function.

- I strongly prefer for only one strategy to work. Either modify in place or use the returned output. I'd guess that the second approach would be a lot slower so I tend to prefer the first solution which would be more inline with general strategy in #11358.
- I dislike adding a second slightly modified version of `apply_along_axis`. Not sure if it would be better to add a switch to the original `apply_along_axis` that toggles allocating the output.

> [...] whether we should update the docstring as proposed here (leaving open the possibility to optimize as suggested in the prior comment) or [...]

I prefer this solution and think it would be worth to further discuss how to avoid the unnecessary allocation in `apply_along_axis`.

File: numpy/lib/arraypad.py [hunk 1]
```diff
@@ -1075,9 +1075,8 @@ def pad(array, pad_width, mode='constant', **kwargs):
     think about with a rank 2 array where the corners of the padded array
     are calculated by using padded values from the first axis.
 
-    The padding function, if used, should return a rank 1 array equal in
-    length to the vector argument with padded values replaced. It has the
-    following signature::
+    The padding function, if used, should modify a rank 1 array in-place. It
+    has the following signature::
```
Comment (lagru) [inline numpy/lib/arraypad.py]: Might I suggest another quick fix to the docstring here? In the explanation of `padding_func`'s arguments the variable `pad_tuple` is used which doesn't appear in the signature. I guess it should be called `iaxis_pad_width` instead? (GitHub wouldn't let me comment in the actual place, line 1087)

Comment (grlee77) [reply numpy/lib/arraypad.py]: good catch. I have made the suggested change

**Comment by grlee77** [comment]
> I dislike adding a second slightly modified version of apply_along_axis. 

Yeah, I agree that it would be better to support it within `apply_along_axis` itself. It is an easy change within the function to allow skipping the buffer creation. The tricky part is if/how to modify the function signature and whether to publicly expose this functionality in the API. It currently has the signature:
```apply_along_axis(func1d, axis, arr, *args, **kwargs)```
where `*args` and `**kwargs` get passed along to `func1d`. This makes it hard to add new optional arguments without potentially breaking existing code that could have potentially been passing arguments of the same name along to `func1d`.

If we don't care to make the functionality public, a hack to avoid modifying the signature could be to check for some "private" keyword such as `_apply_inplace` and pop it from  `kwargs` before passing them along to `func1d`. That doesn't seem the most appealing, though....

**Comment by grlee77** [comment]
It looks like this in-place assumption has been around since at least numpy 1.7 (that was the oldest tag I looked at). Assuming there is agreement to document it as requiring in-place operation, the modified test case here should make sure this is not inadvertently changed going forward.

We can decide about potentially optimizing `apply_along_axis` in a separate issue.

File: numpy/lib/arraypad.py [hunk 2]
```diff
@@ -1147,7 +1146,7 @@ def pad(array, pad_width, mode='constant', **kwargs):
     ...     pad_value = kwargs.get('padder', 10)
     ...     vector[:pad_width[0]] = pad_value
     ...     vector[-pad_width[1]:] = pad_value
-    ...     return vector
+    ...     return
```
Comment (eric-wieser) [inline numpy/lib/arraypad.py]: I'd remove this entirely

Comment (mattip) [reply numpy/lib/arraypad.py]: both here and in the test

Comment (grlee77) [reply numpy/lib/arraypad.py]: sure. I have removed it from both now.

**Comment by mattip** [comment]
Thanks @grlee77, it would be nice to open a follow-on issue to optimize `apply_along_axis`

**Pull request merged**